x-pack/legacy/plugins/ml/public/jobs/new_job_new/common/job_creator/util/general.ts

```diff
@@ -11,32 +11,98 @@ import {
   ML_JOB_AGGREGATION,
   SPARSE_DATA_AGGREGATIONS,
 } from '../../../../../../common/constants/aggregation_types';
-import { EVENT_RATE_FIELD_ID, AggFieldPair } from '../../../../../../common/types/fields';
+import { MLCATEGORY } from '../../../../../../common/constants/field_types';
+import {
+  EVENT_RATE_FIELD_ID,
+  Field,
+  AggFieldPair,
+  mlCategory,
+} from '../../../../../../common/types/fields';
 import { mlJobService } from '../../../../../services/job_service';
 import { JobCreatorType, isMultiMetricJobCreator, isPopulationJobCreator } from '../';
 import { CREATED_BY_LABEL, JOB_TYPE } from './constants';
 
 // populate the detectors with Field and Agg objects loaded from the job capabilities service
-export function getRichDetectors(job: Job, datafeed: Datafeed, advanced: boolean = false) {
+export function getRichDetectors(
+  job: Job,
+  datafeed: Datafeed,
+  scriptFields: Field[],
+  advanced: boolean = false
+) {
   const detectors = advanced ? getDetectorsAdvanced(job, datafeed) : getDetectors(job, datafeed);
+
+  function getFieldById(id: string) {
+    let field = newJobCapsService.getFieldById(id);
+
+    // if no field could be found it may be a pretend field, like mlcategory or a script field
+    if (field === null) {
+      if (id === MLCATEGORY) {
+        field = mlCategory;
+      } else if (scriptFields.length) {
+        field = scriptFields.find(f => f.id === id) || null;
+      }
+    }
+    return field;
+  }
```

Comment:
Nit: Maybe move this one out of this scope e.g. to be able to write tests for it in isolation?